proposals/0000-provenance-qualified-imports.rst

```diff
@@ -0,0 +1,96 @@
+
+Additionally, this proposal allows depdencies on packages hosted outside of a repo or overlay -- such as github, just on the web more generally, or within a filesystem (such as on a shared drive). A use case here would be if I had a package undergoing a major rewrite, and which I did not yet want to release a new version of, but which I wanted to encourage others to take for a test-drive. Rather than asking users to install it directly, or clone and vendor it into their tree, I could now tell them how to use provenance qualification to depend on a particular snapshot of the source repo through altering a few lines in their cabal file.
+
+Proposed Change Specification
+-----------------------------
+
+A) *Provenance Information*
+
+A new optional field, `source-origin` is added to the `InstalledPackageInfo <http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/packages.html#installed-pkg-info>`_ file and in turn to ghc package databases. Tools which register packages to databases may also specify this field. The intended use of this field is to contain a url  that specifies the origin of the package, typically the root of the package repository from which a package was sourced.
+
+B) *Provenance Aliases*
+
+Since urls can be unwieldy, and it is expected that few will be used, it makes sense to allow specification of shorthand aliases to ghc with a new flag. Hence we add a new flag `--source-origin-alias [ALIAS]=[URL]``  that may be specified multiple times. In the event the same alias is specified twice, an error is thrown.
+
+C) *Provenance Specification*
+
+With ``-XPackageImports`` we can specify e.g.
+
+``import "somepackage" Foo.Bar``
+
+Under this proposal we can also specify
+
+``import "[REPO]+somepackage" Foo.Bar``
+
+Where repo is either the url of a repository root or an alias which maps to one.
+
+D) Cabal syntax
+
+Cabal files are extended in two ways.
+
+``source-origin-alises:`` is a top-level property consisting of a list of ``[ALIAS]=[URL]`` aliases. These aliases will be passed to all invocations of `ghc` in the course of a cabal build.
```

Comment:
Woudln’t
```
source-origin-aliases:
  hackage: https://…
  foobar: http://…
```
be more idiomatic?

Is there a way for global aliases in `~/.cabal/config`, with maybe some sensible defaults already there (e.g. `hackage`)?